### DISCUSSION TITLE
When v is compiled as. wasm, does it use GC or autofree memory management by default?

### DISCUSSION THREAD

**Discussion by Jengro777** [vlang/v#19212] [Questions and Answers] [answered]
When compiling v into a wasm executable file using the command `v -b wasm -os wasi xx.v`, the memory management is handled in the following ways:

> 1、It uses the `autofree` mechanism to manage memory. This means that memory is automatically freed when it is no longer needed, reducing the need for manual memory management.
> 
> 2、The` GC` (Garbage Collection) runtime is compiled into wasm along with the code. This allows for automatic memory management and garbage collection within the wasm environment.
> 
> 3、The wasm GC can be utilized using the built-in wasm GC features of runtimes such as wasmtime or wasmer. These runtimes provide additional functionality for managing memory within the wasm environment.

**Comment by JalonSolov** [comment]
@l1mey112 :point_up:

**Comment by l1mey112** [comment, marked as answer]
Currently it uses neither. Just a simple bump allocator able to allocate at possibly at most 64KiBs without crashing. This was a basic allocator created for running the tests and all programs the wasm backend is capable of compiling.

Full code behind `vlib/builtin/wasm/alloc.v`:

```v
@[has_globals]
module builtin

// Shitty `sbrk` basic `malloc` and `free` impl
// TODO: implement pure V `walloc` later

__global g_heap_base = usize(vwasm_heap_base())

// malloc dynamically allocates a `n` bytes block of memory on the heap.
// malloc returns a `byteptr` pointing to the memory address of the allocated space.
// unlike the `calloc` family of functions - malloc will not zero the memory block.
@[unsafe]
pub fn malloc(n isize) &u8 {
	if n <= 0 {
		panic('malloc(n <= 0)')
	}

	res := g_heap_base
	g_heap_base += usize(n)

	return &u8(res)
}

// free allows for manually freeing memory allocated at the address `ptr`.
// currently does not free any memory.
@[unsafe]
pub fn free(ptr voidptr) {
	_ := ptr
}
```
